



<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="An internal resource explaining and documenting lab pipelines">
      
      
        <link rel="canonical" href="https://zamanianlab.org/ZamanianLabDocs/pipelines_server/">
      
      
        <meta name="author" content="Zamanian Lab">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-4.6.0">
    
    
      
        <title>Server Pipelines - Zamanian Lab Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.1b62728e.css">
      
        <link rel="stylesheet" href="../assets/stylesheets/application-palette.a8b3c06d.css">
      
      
        
        
        <meta name="theme-color" content="#3f51b5">
      
    
    
      <script src="../assets/javascripts/modernizr.268332fc.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
    
    
      
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="indigo" data-md-color-accent="teal">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#server-pipelines" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="https://zamanianlab.org/ZamanianLabDocs" title="Zamanian Lab Documentation" class="md-header-nav__button md-logo">
          
            <i class="md-icon"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            <span class="md-header-nav__topic">
              Zamanian Lab Documentation
            </span>
            <span class="md-header-nav__topic">
              
                Server Pipelines
              
            </span>
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
          
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  

<a href="https://github.com/zamanianlab/ZamanianLabDocs/" title="Go to repository" class="md-source" data-md-source="github">
  
    <div class="md-source__icon">
      <svg viewBox="0 0 24 24" width="24" height="24">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    zamanianlab/ZamanianLabDocs
  </div>
</a>
          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main" role="main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="https://zamanianlab.org/ZamanianLabDocs" title="Zamanian Lab Documentation" class="md-nav__button md-logo">
      
        <i class="md-icon"></i>
      
    </a>
    Zamanian Lab Documentation
  </label>
  
    <div class="md-nav__source">
      


  

<a href="https://github.com/zamanianlab/ZamanianLabDocs/" title="Go to repository" class="md-source" data-md-source="github">
  
    <div class="md-source__icon">
      <svg viewBox="0 0 24 24" width="24" height="24">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    zamanianlab/ZamanianLabDocs
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../labmanual/" title="Lab Manual" class="md-nav__link">
      Lab Manual
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../protocols/protocols/" title="Lab Protocols" class="md-nav__link">
      Lab Protocols
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../labsheets/" title="Lab Sheets" class="md-nav__link">
      Lab Sheets
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5">
    
    <label class="md-nav__link" for="nav-5">
      Computational Guide
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-5">
        Computational Guide
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../comp_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../comp_local/" title="Local Environment" class="md-nav__link">
      Local Environment
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../comp_github/" title="GitHub" class="md-nav__link">
      GitHub
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../comp_conda/" title="Conda" class="md-nav__link">
      Conda
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../comp_homebrew/" title="Homebrew" class="md-nav__link">
      Homebrew
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6" checked>
    
    <label class="md-nav__link" for="nav-6">
      Pipelines
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-6">
        Pipelines
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../pipelines_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../pipelines_local/" title="Local Pipelines" class="md-nav__link">
      Local Pipelines
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Server Pipelines
      </label>
    
    <a href="./" title="Server Pipelines" class="md-nav__link md-nav__link--active">
      Server Pipelines
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-lab-storage" class="md-nav__link">
    1. Lab Storage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-center-for-high-throughput-computing-chtc" class="md-nav__link">
    2. Center for High-throughput Computing (CHTC)
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-htc-system" class="md-nav__link">
    The HTC System
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#deploying-pipelines" class="md-nav__link">
    Deploying Pipelines
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-docker" class="md-nav__link">
    2. Docker
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#building-docker-images" class="md-nav__link">
    Building Docker Images
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#testing-docker-pipelines" class="md-nav__link">
    Testing Docker Pipelines
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../editing/" title="Editing" class="md-nav__link">
      Editing
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-lab-storage" class="md-nav__link">
    1. Lab Storage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-center-for-high-throughput-computing-chtc" class="md-nav__link">
    2. Center for High-throughput Computing (CHTC)
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-htc-system" class="md-nav__link">
    The HTC System
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#deploying-pipelines" class="md-nav__link">
    Deploying Pipelines
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-docker" class="md-nav__link">
    2. Docker
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#building-docker-images" class="md-nav__link">
    Building Docker Images
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#testing-docker-pipelines" class="md-nav__link">
    Testing Docker Pipelines
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/zamanianlab/ZamanianLabDocs/edit/master/docs/pipelines_server.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <h1 id="server-pipelines">Server Pipelines</h1>
<p>Our lab has access to powerful computing resources and support through the <a href="http://chtc.cs.wisc.edu/">Center for High-Throughput Computing (CHTC)</a>. We also have access to virtual servers through the UW <a href="https://www.biotech.wisc.edu/services/brc">Bioinformatics Research Center (BRC)</a>, attached to sequencing data generated by the <a href="https://www.biotech.wisc.edu/">UW Biotech Center</a>. Our core bioinformatics and image processing pipelines will be deployed through CHTC servers. All pipelines will be maintained on <a href="http://www.github.com">GitHub</a> and associated with <a href="http://www.docker.com">Docker</a> environments to ensure reproducibility. Many of our pipelines will use <a href="https://www.nextflow.io/">Nextflow</a>.</p>
<p>In general, pipelines will be run in three steps:</p>
<ul>
<li><strong>Staging:</strong> input files will be transferred to the CHTC server from lab storage</li>
<li><strong>Pipeline:</strong> files will be processed using established pipelines</li>
<li><strong>Output:</strong> desired outputs will be transferred from the CHTC server to lab storage</li>
</ul>
<h2 id="1-lab-storage">1. Lab Storage</h2>
<p>Large data sets (raw and processed) will be stored in locations with reliable backup.</p>
<ul>
<li>
<p>BRC (UW Biotech Center)</p>
<p>Sequencing data generated by the UW Biotechnology Center is delivered to their BRC servers. These data can be accessed in <a href="https://www.biotech.wisc.edu/services/brc/data-access">numerous ways</a>.</p>
</li>
<li>
<p>UW ResearchDrive</p>
<p><a href="https://it.wisc.edu/services/researchdrive">ResearchDrive</a> provides 5 TB (expandable) of secure backed-up storage. Lab members will have access to the lab ResearchDrive using their UW net-id and password. Our sequencing data from the BRC is automatically moved into ResearchDrive. We will also use ResearchDrive to store imaging and all other large data types. General instructions on how to connect to and transfer data in and out of ResearchDrive are provided <a href="https://kb.wisc.edu/researchdata/internal/page.php?id=93998">here</a>. Data will be transferred from ResearchDrive to the CHTC for pipeline-based processing. Outputs will be transferred from the CHTC back to ResearchDrive for long-term storage. Subsets of outputs will be transferred into our lab <a href="http://www.box.com">Box</a> account for post-processing, analysis and plotting. Curent ResearchDrive directory structure is shown below:</p>
<div class="highlight"><pre><span></span>/
├── ImageXpress/                          [IX storage]
|   └── raw/                              [Raw exported data]
|   └── metadata/                         [Experiment metadata]    
|   └── data/                             [CHTC-processed data]
|  
└── UWBC-Dropbox/                         [Auto-deposited data from the UWBC]
    └── Bioinformatics Resource Center/   [Sequencing Data]
    └── DNA Sequencing Sanger/            [Sanger Data]
</pre></div>

<p>To connect to ResearchDrive on an Apple Computer,</p>
<div class="highlight"><pre><span></span>Finder &gt; Go &gt; &quot;Connect to server... &gt;
smb://research.drive.wisc.edu/mzamanian
</pre></div>

</li>
<li>
<p>SVM Research Data Storage</p>
<p>SVM PIs will soon have access to secure and backed-up research data storage through the school, stocked initially with 335 TB of storage capacity. Once this storage solution is in place, we will migrate from UW ResearchDrive to the on-premise storage solution.</p>
</li>
</ul>
<h2 id="2-center-for-high-throughput-computing-chtc">2. Center for High-throughput Computing (CHTC)</h2>
<p>Consult official <a href="http://chtc.cs.wisc.edu/">CHTC</a> and <a href="https://research.cs.wisc.edu/htcondor/">HTCondor</a> documentation before getting started. Register for an account using this <a href="http://chtc.cs.wisc.edu/form.shtml">form</a>.</p>
<h3 id="the-htc-system">The HTC System</h3>
<ol>
<li>
<p>Compute Nodes</p>
<p>The CHTC has an extensive set of execute nodes that can be accessed for free. To establish priority access for certain pipelines, our lab has secured a dedicated node that can be accessed using a designated flag.</p>
<ul>
<li>Typical nodes: 20 cores, 128 GB RAM</li>
<li><a href="http://chtc.cs.wisc.edu/high-memory-jobs.shtml">High-memory nodes</a>: e.g., 80 cores, 4 TB RAM</li>
<li>Dedicated lab node: 40 cores, 512 GB RAM, 3.8 TB HD</li>
</ul>
</li>
<li>
<p>Submit nodes</p>
<p>Jobs on the CHTC are deployed from submit nodes. You can <code>ssh</code> into your assigned submit node to run and monitor jobs using your UW net-id and password.</p>
<p><code>ssh {net-id}@submit3.chtc.wisc.edu</code></p>
</li>
<li>
<p>File system</p>
<p>The CHTC does not use a shared file system, but you can request the storage you need for any given job. Each net-id will be associated with a <code>home</code> folder, where you will manage your job submission scripts. Each net-id will also be associated with a <code>staging</code> folder, for transfer of files in and out of the CHTC system.</p>
<div class="highlight"><pre><span></span>/
├── home/{net-id}/              [initial quota: 20 GB, submit script dir]
└── staging/{net-id}/           [initial quota: 200 GB | 1000 files]
    └── input/                  [input dir: unprocessed (raw) data]
    └── output/                 [output dir: processed job outputs]
</pre></div>

</li>
</ol>
<h3 id="deploying-pipelines">Deploying Pipelines</h3>
<ol>
<li>
<p>Staging input data for processing</p>
<p>Transfer a single folder containing your job input files to the staging data directory. You can run transfer commands from the lab file server or BRC server (sequencing data).</p>
<p><code>scp [dir] {net-id}@transfer.chtc.wisc.edu:/staging/{net-id}/data/</code></p>
<p>More typically, you will be <a href="http://chtc.cs.wisc.edu/transfer-data-researchdrive.shtml">transferring directly</a> between ResearchDrive and CHTC. To transfer a directory from ResearchDrive to the CHTC staging input folder:</p>
<p><details>
<summary> ResearchDrive -&gt; CHTC transfer (Click to Expand)</summary>
<div class="highlight"><pre><span></span># log into CHTC staging server and navigate to input folder
ssh {net-id}@transfer.chtc.wisc.edu
cd /staging/{net-id}/input/

# connect to lab ResearchDrive
smbclient -k //research.drive.wisc.edu/mzamanian

# navigate to ResearchDrive dir with raw data (example)
smb: \&gt; cd /ImageXpress/raw/

# turn off prompting and turn on recursive
smb: \&gt; prompt
smb: \&gt; recurse

# transfer raw data folder (example)
smb: \&gt; mget 20200922-p01-NJW_114
</pre></div>
</details></p>
</li>
<li>
<p>Creating job submit scripts</p>
<p>CHTC uses HTCondor for job scheduling. Submission files (.sub) should follow lab conventions and be consistent with the CHTC documentation. An example submit script with annotations is shown below. This submit script (Core_RNAseq-nf.sub) loads a pre-defined <a href="https://hub.docker.com/repository/docker/zamanianlab/chtc-rnaseq">Docker environment</a> and runs a bash executable script (Core_RNAseq-nf.sh) with defined arguments (staged data location).</p>
<p>Other options define standard log files, resource requirements (cpu, memory, and hard disk), and transfer of files in/out of <code>home</code>. Avoid transferring large files in/out of <code>home</code>! We transfer in our large data through <code>/staging/{net-id}/data/</code> and we move job output files to <code>/staging/{net-id}/output/</code> within the job executable script to avoid their transfer to <code>home</code> upon job completion. The only files that should be transferred back to <code>home</code> are small log files.</p>
<p><details>
  <summary>Core_RNAseq-nf.sub (Click to Expand)</summary>
  <div class="highlight"><pre><span></span># Core_RNAseq-nf.sub
# Submit scripts (.sub/.sh) in /home/{net-id}/; Data in /staging/{net-id}/data/$(data)
# Run: condor_submit Core_RNAseq-nf.sub data=191211_AHMMC5DMXX script=Core_RNAseq-nf.sh

# load docker image and request nodes with access to staging directory
universe = docker
docker_image = zamanianlab/chtc-rnaseq:latest

# require execute servers that have large data staging
Requirements = (Target.HasCHTCStaging == true)

# executable (/home/{net-id}/) and arguments
executable = $(script)
arguments = $(data)

# log, error, and st output files
log = $(data)_$(Cluster)_$(Process).log
error = $(data)_$(Cluster)_$(Process).err
output = $(data)_$(Cluster)_$(Process).out

# transfer files in/out of /home/{net-id}/
transfer_input_files =
should_transfer_files = YES
when_to_transfer_output = ON_EXIT

# memory, disk and CPU requests
request_cpus = 4
request_memory = 32GB
request_disk = 20GB

# submit 1 job
queue 1
### END
</pre></div>
</details></p>
<p>The submit script runs the annotated bash script below on the execute server. This pipeline creates <code>data</code>, <code>work</code>, and <code>output</code> dirs in the loaded Docker environment. It transfers the input data from <code>staging</code> into <code>data</code>, clones a GitHub repo (Nextflow pipeline), and runs a Nextflow command. Nextflow uses <code>work</code> for intermediary processing and spits out any files we have marked for retention into <code>output</code>, which gets transferred back to <code>staging</code>. <code>data</code> and <code>work</code> are deleted before job completion.</p>
<p><details>
  <summary>Core_RNAseq-nf.sh (Click to Expand)</summary>
  <div class="highlight"><pre><span></span>#!/bin/bash

# set home () and mk dirs
export HOME=$PWD
mkdir data work output

# transfer data from staging
cp -r /staging/{net-id}/data/$1 data

# clone nextflow git repo
git clone https://github.com/zamanianlab/Core_RNAseq-nf.git

# run nextflow
export NXF_OPTS=&#39;-Xms1g -Xmx4g&#39;
nextflow run Core_RNAseq-nf/WB-pe.nf -w work -c Core_RNAseq-nf/chtc.config --dir $1 --release &quot;WBPS14&quot; --species &quot;brugia_malayi&quot; --prjn &quot;PRJNA10729&quot; --rlen &quot;150&quot;

# rm files you don&#39;t want transferred back to /home/{net-id}
rm -r work
rm -r data

# mv large output files to staging to avoid their transfer back to /home/{net-id}
mv output/$1/ /staging/{net-id}/output/
</pre></div>
</details></p>
</li>
<li>
<p>Submitting and managing jobs</p>
<p>Submit job from submit node using <code>condor_submit</code>,</p>
<p><code>condor_submit Core_RNAseq-nf.sub data=191211_AHMMC5DMXX script=Core_RNAseq-nf.sh</code></p>
<p><details>
  <summary>Other useful commands for monitoring and managing jobs (Click to Expand)</summary>
    <div class="highlight"><pre><span></span># check on job status
  condor_q

# remove a specific job
  condor_rm [job id]

# remove all jobs for user
  condor_rm $USER

# interative shell to running job on remote machine
  condor_ssh_to_job [job id]
  exit
</pre></div>
</details></p>
</li>
<li>
<p>Transferring output data</p>
</li>
</ol>
<h2 id="2-docker">2. Docker</h2>
<p>We will user Docker to establish consistent environments (containers) for our established pipelines. We will maintain Docker images on <a href="https://hub.docker.com/orgs/zamanianlab">Docker Hub</a> under the organization name 'zamanianlab'. These images can be directly loaded from Docker Hub in our CHTC submit scripts. Install <a href="https://docs.docker.com/docker-for-mac/install/">Docker Desktop for Mac</a> and create a Dockerhub account to be associated with our organization Docker Hub (zamanianlab).</p>
<h3 id="building-docker-images">Building Docker Images</h3>
<ol>
<li>
<p>Create a lab Docker Hub repo (e.g., zamanianlab/chtc-rnaseq)</p>
</li>
<li>
<p>Create Dockerfile and auxillary (e.g., yaml) files in a folder</p>
<p>The Dockerfile provides instructions to build a Docker image. In this case, we are starting with the official miniconda Docker image and then installing necessary conda packages into this image. You can search for existing Docker images on <a href="https://hub.docker.com/orgs/zamanianlab">Docker Hub</a> to build on, instead of starting from scratch.</p>
<p><details>
  <summary>Dockerfile (Click to Expand)</summary>
  <div class="highlight"><pre><span></span>FROM continuumio/miniconda3
MAINTAINER mzamanian@wisc.edu

# install (nf tracing)
RUN apt-get update &amp;&amp; apt-get install -y procps

# install conda packages
COPY conda_env.yml .
RUN \
   conda env update -n root -f conda_env.yml \
&amp;&amp; conda clean -a
</pre></div>
</details></p>
<p>yml file containing <code>conda</code> packages to be installed. You can search for packages on <a href="https://anaconda.org/">Anaconda cloud</a>.</p>
<p><details>
  <summary>conda_env.yml (Click to Expand)</summary>
  <div class="highlight"><pre><span></span>conda_env.yaml
  name: rnaseq-nf

  channels:
    - bioconda
    - conda-forge
    - defaults

  dependencies:
    - python=3.8.5
    - nextflow=20.07.1
    - bwa=0.7.17
    - hisat2=2.2.1
    - stringtie=2.1.2
    - fastqc=0.11.9
    - multiQC=1.9
    - fastp=0.20.1
    - bedtools=2.29.2
    - bedops=2.4.39
    - sambamba=0.7.0
    - samtools=1.9
    - picard=2.20.6
    - bcftools=1.9
    - snpeff=4.3.1t
    - mrbayes=3.2.7
    - trimal=1.4.1
    - mafft=7.471
    - muscle=3.8.1551
    - seqtk=1.3
    - raxml=8.2.12
</pre></div>
</details></p>
</li>
<li>
<p>Build Docker image</p>
<div class="highlight"><pre><span></span>cd [/path/to/Dockerfile]
docker build -t zamanianlab/chtc-rnaseq .
</pre></div>

</li>
<li>
<p>Test Docker image interactively</p>
<div class="highlight"><pre><span></span>docker run -it --rm=TRUE zamanianlab/chtc-rnaseq /bin/bash
ctrl+D to exit
</pre></div>

</li>
<li>
<p>Push Docker image to Docker Hub</p>
<div class="highlight"><pre><span></span>docker push zamanianlab/chtc-rnaseq
</pre></div>

<p><details>
  <summary>Some useful Docker commands (Click to Expand)</summary>
    <div class="highlight"><pre><span></span># list docker images
  docker image ls (= docker images)

# remove images
  docker rmi [image]

## remove all docker containers
# run first because images are attached to containers
  docker rm -f $(docker ps -a -q)
# remove every Docker image
  docker rmi -f $(docker images -q)
</pre></div>
</details></p>
</li>
</ol>
<h3 id="testing-docker-pipelines">Testing Docker Pipelines</h3>
                
                  
                
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../pipelines_local/" title="Local Pipelines" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Local Pipelines
              </span>
            </div>
          </a>
        
        
          <a href="../editing/" title="Editing" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Editing
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
  <div class="md-footer-social">
    <link rel="stylesheet" href="../assets/fonts/font-awesome.css">
    
      <a href="https://github.com/zamanianlab" class="md-footer-social__link fa fa-"></a>
    
      <a href="https://twitter.com/zamanian_" class="md-footer-social__link fa fa-"></a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.808e90bb.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:".."}})</script>
      
        <script src="../javascripts/extra.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      
    
  </body>
</html>